### PR TITLE
Cognito pool domain creation issue

### PR DESCRIPTION
**Pull request by Biplob14**
#Fixes 

- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [ ] New feature
- [x] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?
- On branch deployment cognito domain will be created upon branch name.
- Branch name special characters are replaced as necessary to follow cognito naming convention. 
- Only on production cognito custom domain will be created with root domain A name.

## Related Issues/PRs (optional)
Complexity on cognito domain creation #175 

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [x] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [x] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)
Please check if there is any logical correction is necessary on resource `aws_cognito_user_pool_domain`. 

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [ ] No risk
- [x] Low
- [ ] High

### CODE REVIEW

File: iac/terraform/aws/cognito/cognito.tf [hunk 1]
```diff
@@ -57,6 +57,35 @@ resource "aws_cognito_user_pool" "user_pool" {
...
+resource "aws_cognito_user_pool_domain" "set_custom_domain_prod" {
+  count = var.environment == module.environments.PRODUCTION ? 1 : 0
+
+  domain          = "${var.bloodconnect_domain}"
+  user_pool_id    = aws_cognito_user_pool.user_pool.id
+  certificate_arn = var.acm_certificate_arn
+}
+
+resource "aws_cognito_user_pool_domain" "set_custom_domain" {
+  count = var.environment == module.environments.PRODUCTION ? 0 : 1
```
Comment (sakibstark11) [inline iac/terraform/aws/cognito/cognito.tf:69]: maybe just move this into a local block that says

```hcl
locals {
    isProduction = var.environment == module.environments.PRODUCTION ? 0 : 1
}
```

File: iac/terraform/aws/cognito/cognito.tf [hunk 2]
```diff
@@ -57,6 +57,35 @@ resource "aws_cognito_user_pool" "user_pool" {
   }
 }
 
+resource "aws_cognito_user_pool_domain" "set_custom_domain_prod" {
+  count = var.environment == module.environments.PRODUCTION ? 1 : 0
+
+  domain          = "${var.bloodconnect_domain}"
```
Comment (sakibstark11) [inline iac/terraform/aws/cognito/cognito.tf:63]: no need for the quotes

File: iac/terraform/aws/cognito/cognito.tf [hunk 3]
```diff
@@ -57,6 +57,35 @@ resource "aws_cognito_user_pool" "user_pool" {
...
+
+  domain          = replace(replace(replace(replace("${var.environment}", "aws", ""), "amazon", ""), "cognito", ""), "-", "")
+  user_pool_id    = aws_cognito_user_pool.user_pool.id
+}
+
+resource "aws_route53_record" "cognito_user_pool_custom_domain" {
+  count = var.environment == module.environments.PRODUCTION ? 1 : 0
+
+  name    = aws_cognito_user_pool_domain.set_custom_domain_prod[0].domain
+  type    = "A"
```
Comment (sakibstark11) [inline iac/terraform/aws/cognito/cognito.tf:79]: are you sure its a A type? its not an ip address, A records are IP addresses

File: iac/terraform/aws/cognito/cognito.tf (hunk 1 above)
Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf:69]: Nice idea. Moving it

File: iac/terraform/aws/cognito/cognito.tf (hunk 3 above)
Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf:79]: Yes , according to the [Custom Cognito domain](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cognito_user_pool_domain#custom-cognito-domain)

**Review by KhirulIslam — DISMISSED** [review]

**Review by shuaibvm — COMMENTED** [review]
we are introducing a var in terraform by the name `locals`, where we have `isProduction` and `isNonProduction`, where we set `1` or `0` based on `var.environment`.

firstly, the name in itself seems a bit confusing to me, as in `locals.isProduction` and `locals.isNonProduction`.. locals should not have anything called production and non production in my opinion. or maybe it is appropriate in this context, which i am not aware of. please share your opinion on this.

secondly, variables starting with `is` ideally should return a boolean, and not a number. this basically returns a number that is set to count. maybe apply the logic directly at `count` if this value is not required elsewhere? or maybe use a name that does not start with `is` in my opinion?

please feel free to disagree and put in your thoughts here.

**Review by shuaibq3 — COMMENTED** [review]
we are introducing a var in terraform by the name `locals`, where we have `isProduction` and `isNonProduction`, where we set `1` or `0` based on var.environment.

firstly, the name in itself seems a bit confusing to me, as in `locals.isProduction` and `locals.isNonProduction`.. locals should not have anything called production and non production in my opinion. or maybe it is appropriate in this context, which i am not aware of. please share your opinion on this.

secondly, variables starting with `is` ideally should return a boolean, and not a number. this basically returns a number that is set to `count`. maybe apply the logic directly at count if this value is not required elsewhere? or maybe use a name that does not start with `is`?

please feel free to disagree and put in your thoughts here.

File: iac/terraform/aws/cognito/cognito.tf (hunk 3 above)
Comment (sakibstark11) [reply iac/terraform/aws/cognito/cognito.tf:79]: the link doesn't tell me anything

Comment (sakibstark11) [reply iac/terraform/aws/cognito/cognito.tf:79]: i feel like this will be a CNAME

**Comment by sakibstark11** [comment]
> we are introducing a var in terraform by the name `locals`, where we have `isProduction` and `isNonProduction`, where we set `1` or `0` based on var.environment.
> 
> firstly, the name in itself seems a bit confusing to me, as in `locals.isProduction` and `locals.isNonProduction`.. locals should not have anything called production and non production in my opinion. or maybe it is appropriate in this context, which i am not aware of. please share your opinion on this.
> 
> secondly, variables starting with `is` ideally should return a boolean, and not a number. this basically returns a number that is set to `count`. maybe apply the logic directly at count if this value is not required elsewhere? or maybe use a name that does not start with `is`?
> 
> please feel free to disagree and put in your thoughts here.

I think locals are meant to be used to reduce duplication. So, having a variable that identifies production is a good thing to have given we have special cases where we only do things if its production. I do however agree that maybe calling it isProduction is not a good idea. Just plain production is probably better. When I suggested this, I used it as an example and I suppose Biplop took it literally.

**Comment by sakibstark11** [comment]
> we are introducing a var in terraform by the name `locals`, where we have `isProduction` and `isNonProduction`, where we set `1` or `0` based on var.environment.
> 
> firstly, the name in itself seems a bit confusing to me, as in `locals.isProduction` and `locals.isNonProduction`.. locals should not have anything called production and non production in my opinion. or maybe it is appropriate in this context, which i am not aware of. please share your opinion on this.
> 
> secondly, variables starting with `is` ideally should return a boolean, and not a number. this basically returns a number that is set to `count`. maybe apply the logic directly at count if this value is not required elsewhere? or maybe use a name that does not start with `is`?
> 
> please feel free to disagree and put in your thoughts here.

Also using count as a boolean in terraform is a common practice. Since terraform itself doesn't have a representation of boolean values.

**Comment by Biplob14** [comment]
> > we are introducing a var in terraform by the name `locals`, where we have `isProduction` and `isNonProduction`, where we set `1` or `0` based on var.environment.
> > firstly, the name in itself seems a bit confusing to me, as in `locals.isProduction` and `locals.isNonProduction`.. locals should not have anything called production and non production in my opinion. or maybe it is appropriate in this context, which i am not aware of. please share your opinion on this.
> > secondly, variables starting with `is` ideally should return a boolean, and not a number. this basically returns a number that is set to `count`. maybe apply the logic directly at count if this value is not required elsewhere? or maybe use a name that does not start with `is`?
> > please feel free to disagree and put in your thoughts here.
> 
> I think locals are meant to be used to reduce duplication. So, having a variable that identifies production is a good thing to have given we have special cases where we only do things if its production. I do however agree that maybe calling it isProduction is not a good idea. Just plain production is probably better. When I suggested this, I used it as an example and I suppose Biplop took it literally.

As we are declaring an operation only in production environment and some other operation in all other environment, I thought this name as most expressing in this situation.

**Comment by Biplob14** [comment]
> we are introducing a var in terraform by the name `locals`, where we have `isProduction` and `isNonProduction`, where we set `1` or `0` based on var.environment.
> 
> firstly, the name in itself seems a bit confusing to me, as in `locals.isProduction` and `locals.isNonProduction`.. locals should not have anything called production and non production in my opinion. or maybe it is appropriate in this context, which i am not aware of. please share your opinion on this.
> 
> secondly, variables starting with `is` ideally should return a boolean, and not a number. this basically returns a number that is set to `count`. maybe apply the logic directly at count if this value is not required elsewhere? or maybe use a name that does not start with `is`?
> 
> please feel free to disagree and put in your thoughts here.

We are using `isProduction` in kind of a boolean situation here. Though it may create confusion and I agree with your openion.

File: iac/terraform/aws/cognito/cognito.tf (hunk 3 above)
Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf:116]: Theoretically it should be  `CNAME`  as it's not directly pointing to any `IP address`. I'll check if any issue occurs or not.

**Review by sakibstark11 — APPROVED** [review]

**Review by KhirulIslam — APPROVED** [review]

**Pull request merged**